### PR TITLE
Pause rework m600

### PR DESCRIPTION
**Pull request by RadekVana**
BFW-768

### CODE REVIEW

File: src/common/marlin_server.hpp [hunk 1]
```diff
@@ -13,7 +13,13 @@ void fsm_create(ClinetFSM type, uint8_t data);
...
 void fsm_destroy(ClinetFSM type);
 //notify all clients to change state of finit statemachine, must match fsm_change_t signature
-void fsm_change(ClinetFSM type, uint8_t phase, uint8_t progress_tot, uint8_t progress);
+//can be called inside while, notification is send only when is different from previous one
+void _fsm_change(ClinetFSM type, uint8_t phase, uint8_t progress_tot, uint8_t progress);
+
+template <class T>
+void fsm_change(ClinetFSM type, T phase, uint8_t progress_tot, uint8_t progress) {
+    _fsm_change(type, GetPhaseIndex(phase), progress_tot, progress);
+}
```
Comment (DRracer) [inline src/common/marlin_server.hpp:22]: How many phases do we have in FW? Is GetPhaseIndex marked as constexpr?

File: src/marlin_stubs/pause/M701_M702.cpp [hunk 2]
```diff
@@ -79,12 +80,32 @@ static void load_unload(LoadUnloadMode type, load_unload_fnc f_load_unload, uint
 
 /**
  * Load filament special code
+ * filament type must be last parmaeter
```
Comment (DRracer) [inline src/marlin_stubs/pause/M701_M702.cpp:83]: typo: parmaeter

File: src/marlin_stubs/pause/pause.cpp [hunk 3]
```diff
@@ -175,6 +137,28 @@ void plan_pause_e_move_notify_progress(const float &length, const feedRate_t &fr
     }
 }
 
+//todo move
```
Comment (DRracer) [inline src/marlin_stubs/pause/pause.cpp:140]: make a src/common/RAII.h and place it there

File: src/marlin_stubs/pause/pause.cpp [hunk 4]
```diff
@@ -189,41 +173,83 @@ void plan_pause_e_move_notify_progress(const float &length, const feedRate_t &fr
...
-        planner.settings.retract_acceleration = saved_acceleration;
     }
 
     if (purge_length > 0) {
         Response response;
         do {
             // Extrude filament to get into hotend
             do_pause_e_move_notify_progress(purge_length, ADVANCED_PAUSE_PURGE_FEEDRATE, PhasesLoadUnload::Purging, 70, 99);
-            fsm_change(ClinetFSM::Load_unload, GetPhaseIndex(PhasesLoadUnload::IsColor), 99, 0);
+            fsm_change(ClinetFSM::Load_unload, PhasesLoadUnload::IsColor, 99, 0);
```
Comment (DRracer) [inline src/marlin_stubs/pause/pause.cpp:252]: typo: ClinetFSM

File: src/marlin_stubs/pause/pause.cpp [hunk 5]
```diff
@@ -276,7 +302,7 @@ bool unload_filament(const float &unload_length, const bool show_lcd /*=false*/,
     constexpr size_t ramUnloadSeqSize = sizeof(ramUnloadSeq) / sizeof(RamUnloadSeqItem);
 
     //cannot draw progress in plan_pause_e_move, so just change phase to ramming
-    fsm_change(ClinetFSM::Load_unload, GetPhaseIndex(PhasesLoadUnload::Ramming), 50, 0);
+    fsm_change(ClinetFSM::Load_unload, PhasesLoadUnload::Ramming, 50, 0);
```
Comment (DRracer) [inline src/marlin_stubs/pause/pause.cpp:305]: Why isn't fsm an object (class) encapsulating all its operations?

File: src/common/marlin_server.hpp (hunk 1 above)
Comment (RadekVana) [reply src/common/marlin_server.hpp:22]: Load dialog has 17 phases
other dialogs are not refactored yet

Comment (RadekVana) [reply src/common/marlin_server.hpp:22]: yes, it is

File: src/marlin_stubs/pause/pause.cpp (hunk 3 above)
Comment (RadekVana) [reply src/marlin_stubs/pause/pause.cpp:140]: fixed

File: src/marlin_stubs/pause/M701_M702.cpp (hunk 2 above)
Comment (RadekVana) [reply src/marlin_stubs/pause/M701_M702.cpp:83]: fixed

File: src/marlin_stubs/pause/pause.cpp (hunk 5 above)
Comment (RadekVana) [reply src/marlin_stubs/pause/pause.cpp:305]: because marlin client is written in C

File: src/marlin_stubs/pause/pause.cpp (hunk 4 above)
Comment (RadekVana) [reply src/marlin_stubs/pause/pause.cpp]: fixed

**Pull request merged**